package/metrics_daemon/metrics_daemon/src/metrics_daemon.c

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+#define _GNU_SOURCE
+#include <getopt.h>
+#include <string.h>
+#include <unistd.h>
+#include <ftw.h>
+#include <json-c/json.h>
+#include <string.h>
+#include <libpiksi/logging.h>
+#include <libpiksi/settings.h>
+#include <libpiksi/util.h>
+#include <libpiksi/settings.h>
+#include "sbp.h"
+#include "metrics_daemon.h"
+#define PROGRAM_NAME "metrics_daemon"
+
+#define METRICS_ROOT_DIRECTORY "/var/log/metrics"
+#define METRICS_OUTPUT_FILENAME "/var/log/metrics.json"
+#define METRICS_USAGE_UPDATE_INTERVAL_MS (1000u)
+
+#define MAX_FOLDER_NAME_LENGTH 64
+static json_object *jobj_cur = NULL;
+json_object *jobj_root = NULL;
+char *root_name = NULL;
+unsigned int root_length = 0;
+static bool first_folder = true;
+static const char *target_file = METRICS_OUTPUT_FILENAME;
+static const char *metrics_path = METRICS_ROOT_DIRECTORY;
+static bool enable_log_to_file = false;
+static unsigned int metrics_update_interval = 1;
+
+
+/**
+ * @brief function updates json to file
+ *
+ * @param root       pointer to the root of json object
+ * @param file_name  file to write to
+ */
+static bool write_json_to_file(struct json_object *root, const char *file_path)
+{
+  return file_append_string(file_path, json_object_to_json_string(root));
+}
+
+/**
+ * @brief static function that returns file name from path
+ *
+ * @param  str        file path in ASCII
+ * @return file_name  file name
+ */
+char *extract_filename(const char *str)
+{
+  return basename((char *)str);
+}
+
+
+/**
+ * @brief function that use the full file path loop through the json object tree
+ * to update values
+ *
+ * @param process_path file path in ASCII
+ * @param root         root name, this is used to pair with the root of the json
+ * object tree
+ * @param root_len     number of chars for the root name
+ * @return json object node that needs to be updated
+ */
+struct json_object *loop_through_folder_name(const char *process_path,
+                                             const char *root,
+                                             unsigned int root_len)
+{
+  char slash = '/';
+  const char *start_ptr = NULL;
+  bool found_root = false;
+  start_ptr = process_path;
+  unsigned int str_len = (unsigned int)strlen(process_path);
+  if (start_ptr == NULL) {
```

Comment:
The only time this would fail is if we call `loop_through_folder_name` with `process_path` as `NULL`, in which case `strlen(process_path)` would crash.  Seems this would be more correct:

```c
if (process_path == NULL) return jobj_root;
start_ptr = process_path;
unsigned int str_len = (unsigned int)strlen(process_path);
...
```